firecracker-control/local.go

```diff
@@ -81,263 +67,251 @@ func newLocal(ic *plugin.InitContext) (*local, error) {
-	}
+	return fccontrolTtrpc.NewFirecrackerClient(ttrpc.NewClient(conn)), nil
+}
 
-	for _, drive := range req.GetAdditionalDrives() {
-		driveBuilder = addDriveFromProto(driveBuilder, drive)
+// StopVM stops running VM instance by VM ID
+func (s *local) StopVM(requestCtx context.Context, req *proto.StopVMRequest) (*empty.Empty, error) {
+	client, err := s.shimFirecrackerClient(requestCtx, req.VMID)
+	if err != nil {
+		return nil, err
 	}
 
-	// TODO: Reserve fake drives here (https://github.com/firecracker-microvm/firecracker-containerd/pull/154)
-
-	cfg.Drives = driveBuilder.Build()
-
-	// Setup network interfaces
-
-	for _, ni := range req.GetNetworkInterfaces() {
-		cfg.NetworkInterfaces = append(cfg.NetworkInterfaces, networkConfigFromProto(ni))
+	resp, err := client.StopVM(requestCtx, req)
+	if err != nil {
+		err = errors.Wrap(err, "shim client failed to stop VM")
+		s.logger.WithError(err).Error()
+		return nil, err
 	}
 
-	return cfg, nil
+	return resp, err
 }
 
-// startMachine attempts to start an instance within timeout and runs a goroutine to monitor VM exit event
-func (s *local) startMachine(ctx context.Context, id string, machine machine) error {
-	logger := log.G(ctx).WithField("vm_id", id)
-
-	logger.Info("starting VM instance")
-
-	ctx, cancel := context.WithTimeout(ctx, firecrackerStartTimeout)
-	defer cancel()
+func (s *local) GetVMInfo(requestCtx context.Context, req *proto.GetVMInfoRequest) (*proto.GetVMInfoResponse, error) {
+	client, err := s.shimFirecrackerClient(requestCtx, req.VMID)
+	if err != nil {
+		return nil, err
+	}
 
-	ns, err := namespaces.NamespaceRequired(ctx)
+	resp, err := client.GetVMInfo(requestCtx, req)
 	if err != nil {
-		return err
+		err = errors.Wrap(err, "shim client failed to get vm info")
+		s.logger.WithError(err).Error()
+		return nil, err
 	}
 
-	logger.Debugf("using namespace: %s", ns)
+	return resp, nil
+}
 
-	if err := machine.Start(ctx); err != nil {
-		logger.WithError(err).Error("failed to start the VM")
-		return err
+// SetVMMetadata sets Firecracker instance metadata
+func (s *local) SetVMMetadata(requestCtx context.Context, req *proto.SetVMMetadataRequest) (*empty.Empty, error) {
+	client, err := s.shimFirecrackerClient(requestCtx, req.VMID)
+	if err != nil {
+		return nil, err
 	}
 
-	if err := s.publisher.Publish(ctx, startEventName, &proto.VMStart{VMID: id}); err != nil {
-		logger.WithError(err).Error("failed to publish start event")
-		return err
+	resp, err := client.SetVMMetadata(requestCtx, req)
+	if err != nil {
+		err = errors.Wrap(err, "shim client failed to set vm info")
+		s.logger.WithError(err).Error()
+		return nil, err
 	}
 
-	go func() {
-		ctx := namespaces.WithNamespace(context.Background(), ns)
+	return resp, nil
+}
 
-		if err := machine.Wait(ctx); err != nil {
-			logger.WithError(err).Error("failed to wait the VM")
-		}
+func (s *local) newShim(ns, vmID, containerdAddress string, shimSocket *net.UnixListener, fcSocket *net.UnixListener) (*exec.Cmd, error) {
```

Comment:
Well it definitely started out from their example shim implementation a while back, but it's been modified more and more as time has gone on. Right now, it's still somewhat recognizable but has diverged a fair bit: https://github.com/containerd/containerd/blob/master/runtime/v2/runc/v2/service.go#L117

I'd say in general our entire original shim implementation copied quite a bit of their runtime v2 code at the time (besides just this newShim function). While I *suspect* that's what the containerd authors intended for others to do, I have no idea whether it would be appropriate to attribute to them or not. Let me know what you think and how to attribute them if you think that makes sense.

As for upstream changes, the best I can think of would be if they provided a default "newShimCmd" that we could get and then modify with our customizations, which would let us de-dupe the `cmd.Args` and `cmd.SysProcAttr` lines.

However, that really only saves a couple lines. I'm not convinced it would actually provide much benefit.